File: project_euler/problem_252/solution.py
"""
Project Euler Problem 252: Convex Holes

Find the largest area convex hole among 500 LCG-generated points.
A convex hole is a convex polygon whose vertices are from the point set
and whose interior contains no other points from the set.

Algorithm: For each anchor point p0, sort remaining points by polar angle,
then DP on (last two vertices) tracking first vertex for closure checks.
Empty-triangle tests use precomputed bitsets.

Answer: 104924.0
"""

import math
from functools import cmp_to_key


def cross(O, A, B):
    return (A[0] - O[0]) * (B[1] - O[1]) - (A[1] - O[1]) * (B[0] - O[0])


def solve():
    # Generate points
    s = 290797
    t = []
    for _ in range(1000):
        s = (s * s) % 50515093
        t.append(s % 2000 - 1000)
    pts = [(t[2 * k], t[2 * k + 1]) for k in range(500)]
    n = len(pts)

    WORDS = (n + 63) // 64

    # Precompute bitsets: B[i][j] = set of points strictly left of edge i->j
    B = [[[0] * WORDS for _ in range(n)] for _ in range(n)]
    for i in range(n):
        for j in range(n):
            if i == j:
                continue
            for p in range(n):
                if p == i or p == j:
                    continue
                if cross(pts[i], pts[j], pts[p]) > 0:
                    B[i][j][p // 64] |= 1 << (p % 64)

    def triangle_empty(a, b, c):
        for w in range(WORDS):
            if B[a][b][w] & B[b][c][w] & B[c][a][w]:
                return False
        return True

    best = 0

    for p0 in range(n):
        idx = [i for i in range(n) if i != p0]

        def cmp(a, b):
            ax = pts[a][0] - pts[p0][0]
            ay = pts[a][1] - pts[p0][1]
            bx = pts[b][0] - pts[p0][0]
            by_ = pts[b][1] - pts[p0][1]
            qa = 0 if (ay > 0 or (ay == 0 and ax > 0)) else 1
            qb = 0 if (by_ > 0 or (by_ == 0 and bx > 0)) else 1
            if qa != qb:
                return -1 if qa < qb else 1
            cr = ax * by_ - ay * bx
            if cr != 0:
                return 1 if cr > 0 else -1
            da = ax * ax + ay * ay
            db = bx * bx + by_ * by_
            if da != db:
                return -1 if da < db else 1
            return 0

        idx.sort(key=cmp_to_key(cmp))
        m = len(idx)
        dp = {}
        fv = {}

        # Base: triangles (p0, j, k)
        for j in range(m):
            for k in range(j + 1, m):
                cr = cross(pts[p0], pts[idx[j]], pts[idx[k]])
                if cr <= 0:
                    continue
                if not triangle_empty(p0, idx[j], idx[k]):
                    continue
                dp[(j, k)] = cr
                fv[(j, k)] = j
                best = max(best, cr)

        # Extend
        for k in range(m):
            for j in range(k):
                if (j, k) not in dp:
                    continue
                f = fv[(j, k)]
                area_jk = dp[(j, k)]
                for l in range(k + 1, m):
                    cr1 = cross(pts[idx[j]], pts[idx[k]], pts[idx[l]])
                    if cr1 <= 0:
                        continue
                    cr2 = cross(pts[p0], pts[idx[k]], pts[idx[l]])
                    if cr2 <= 0:
                        continue
                    if not triangle_empty(p0, idx[k], idx[l]):
                        continue
                    if cross(pts[p0], pts[idx[f]], pts[idx[l]]) <= 0:
                        continue
                    new_area = area_jk + cr2
                    if (k, l) not in dp or new_area > dp[(k, l)]:
                        dp[(k, l)] = new_area
                        fv[(k, l)] = f
                    if (cross(pts[idx[k]], pts[idx[l]], pts[p0]) > 0
                            and cross(pts[idx[l]], pts[p0], pts[idx[f]]) > 0):
                        best = max(best, new_area)

    print(f"{best / 2.0:.1f}")


if __name__ == "__main__":
    solve()

File: project_euler/problem_252/solution.cpp
/*
 * Project Euler Problem 252: Convex Holes
 *
 * Find the largest area convex hole among 500 LCG-generated points.
 * A convex hole is a convex polygon whose vertices are from the point set
 * and whose interior contains no other points.
 *
 * Algorithm: For each anchor p0, sort remaining points by polar angle,
 * then DP on (last two vertices) tracking the first vertex for closure.
 * Empty-triangle tests use precomputed bitsets (64-bit words).
 *
 * Answer: 104924.0
 */

#include <bits/stdc++.h>
using namespace std;

typedef long long ll;
typedef pair<ll, ll> pll;

ll cross(pll O, pll A, pll B) {
    return (A.first - O.first) * (B.second - O.second)
         - (A.second - O.second) * (B.first - O.first);
}

int main() {
    // Generate points
    vector<pll> pts;
    ll s = 290797;
    vector<ll> t;
    for (int i = 0; i < 1000; i++) {
        s = (s * s) % 50515093LL;
        t.push_back(s % 2000 - 1000);
    }
    for (int k = 0; k < 500; k++)
        pts.push_back({t[2 * k], t[2 * k + 1]});

    int n = pts.size();
    const int WORDS = (n + 63) / 64;

    // Precompute bitsets: B[i][j] = points strictly left of directed edge i->j
    vector<vector<vector<uint64_t>>> B(n, vector<vector<uint64_t>>(n, vector<uint64_t>(WORDS, 0)));
    for (int i = 0; i < n; i++)
        for (int j = 0; j < n; j++) {
            if (i == j) continue;
            for (int p = 0; p < n; p++) {
                if (p == i || p == j) continue;
                if (cross(pts[i], pts[j], pts[p]) > 0)
                    B[i][j][p / 64] |= (1ULL << (p % 64));
            }
        }

    auto triangle_empty = [&](int a, int b, int c) -> bool {
        for (int w = 0; w < WORDS; w++)
            if (B[a][b][w] & B[b][c][w] & B[c][a][w]) return false;
        return true;
    };

    ll best = 0;

    for (int p0 = 0; p0 < n; p0++) {
        vector<int> idx;
        for (int i = 0; i < n; i++)
            if (i != p0) idx.push_back(i);

        sort(idx.begin(), idx.end(), [&](int a, int b) {
            ll ax = pts[a].first - pts[p0].first, ay = pts[a].second - pts[p0].second;
            ll bx = pts[b].first - pts[p0].first, by = pts[b].second - pts[p0].second;
            int qa = (ay > 0 || (ay == 0 && ax > 0)) ? 0 : 1;
            int qb = (by > 0 || (by == 0 && bx > 0)) ? 0 : 1;
            if (qa != qb) return qa < qb;
            ll cr = ax * by - ay * bx;
            if (cr != 0) return cr > 0;
            return ax * ax + ay * ay < bx * bx + by * by;
        });

        int m = idx.size();
        vector<int> first_vtx(m * m, -1);
        auto FV = [&](int j, int k) -> int& { return first_vtx[j * m + k]; };
        vector<ll> dp(m * m, -1);
        auto DP = [&](int j, int k) -> ll& { return dp[j * m + k]; };

        // Base triangles
        for (int j = 0; j < m; j++)
            for (int k2 = j + 1; k2 < m; k2++) {
                ll cr = cross(pts[p0], pts[idx[j]], pts[idx[k2]]);
                if (cr <= 0) continue;
                if (!triangle_empty(p0, idx[j], idx[k2])) continue;
                DP(j, k2) = cr;
                FV(j, k2) = j;
                best = max(best, cr);
            }

        // Extend
        for (int k2 = 0; k2 < m; k2++)
            for (int j = 0; j < k2; j++) {
                if (DP(j, k2) < 0) continue;
                int f = FV(j, k2);
                for (int l = k2 + 1; l < m; l++) {
                    ll cr1 = cross(pts[idx[j]], pts[idx[k2]], pts[idx[l]]);
                    if (cr1 <= 0) continue;
                    ll cr2 = cross(pts[p0], pts[idx[k2]], pts[idx[l]]);
                    if (cr2 <= 0) continue;
                    if (!triangle_empty(p0, idx[k2], idx[l])) continue;
                    if (cross(pts[p0], pts[idx[f]], pts[idx[l]]) <= 0) continue;

                    ll new_area = DP(j, k2) + cr2;
                    if (new_area > DP(k2, l)) {
                        DP(k2, l) = new_area;
                        FV(k2, l) = f;
                    }
                    if (cross(pts[idx[k2]], pts[idx[l]], pts[p0]) > 0 &&
                        cross(pts[idx[l]], pts[p0], pts[idx[f]]) > 0)
                        best = max(best, new_area);
                }
            }
    }

    printf("%.1f\n", best / 2.0);
    return 0;
}
